File: web/src/links.ts
// Extract bare URLs and `[text](url)` links from descriptions.

// A link found in a description, ready to render.
export interface Link {
  url: string;
  // Explicit Markdown text or a non-empty label derived from the URL.
  label: string;
  // Used by compact affordances that distinguish authored text from fallback.
  explicit: boolean;
}

// Append a URL as its own final paragraph without rewriting any authored text.
// Descriptions are plain text/Markdown, so two line breaks are the smallest
// separator that stays a separate paragraph in both representations.
export function appendLinkToDescription(description: string, url: string): string {
  if (description === "") return url;
  const separator = description.endsWith("\n\n") ? "" : description.endsWith("\n") ? "\n" : "\n\n";
  return `${description}${separator}${url}`;
}

function countChar(s: string, ch: string): number {
  let n = 0;
  for (const c of s) if (c === ch) n++;
  return n;
}

// Strip sentence punctuation while preserving balanced URL parentheses.
export function trimUrl(url: string): string {
  let end = url.length;
  while (end > 0) {
    const ch = url[end - 1]!;
    if (!")].,;!?'\"".includes(ch)) break;
    if (ch === ")" && countChar(url.slice(0, end), "(") >= countChar(url.slice(0, end), ")")) break;
    end--;
  }
  return url.slice(0, end);
}

// Match Markdown first so its target is not emitted again as a bare URL.
// Labels are single-line; targets allow one nested parenthesis pair.
// The http(s) prefix is load-bearing: it is what keeps javascript:/data:
// targets from ever reaching an href.
const LINK_RE = /\[([^\]\n]*)\]\((https?:\/\/(?:[^\s()]|\([^\s()]*\))+)\)|https?:\/\/[^\s]+/g;

// Return distinct HTTP(S) links in first-occurrence order. An explicit label
// replaces a derived one for the same URL without changing that order.
export function extractLinks(text: string): Link[] {
  const byUrl = new Map<string, Link>();
  for (const m of text.matchAll(LINK_RE)) {
    // Markdown targets are group 2; bare URLs need punctuation trimming.
    const url = m[2] ?? trimUrl(m[0]);
    const written = m[1]?.trim();
    const link: Link = written
      ? { url, label: written, explicit: true }
      : { url, label: linkLabel(url), explicit: false };
    const seen = byUrl.get(url);
    // Updating an existing Map key preserves its insertion position.
    if (!seen || (link.explicit && !seen.explicit)) byUrl.set(url, link);
  }
  return [...byUrl.values()];
}

// linkLabel gives a chip a short, human-readable name: a Jira-style issue key
// when the URL is a ".../browse/KEY-123" link, otherwise the host plus the
// last path segment (so two links to the same host stay distinguishable).
export function linkLabel(url: string): string {
  try {
    const u = new URL(url);
    const jira = u.pathname.match(/\/browse\/([A-Z0-9]+-\d+)/);
    if (jira) return jira[1]!;
    const segments = u.pathname.split("/").filter(Boolean);
    const last = segments[segments.length - 1];
    return last ? `${u.hostname}/${last}` : u.hostname;
  } catch {
    return url;
  }
}

File: internal/sheet/links.go
package sheet

import (
	"net/url"
	"regexp"
	"strings"
)

// Descriptions carry links; there is no link field. This is a port of
// web/src/links.ts — Markdown first so its target is not re-emitted as a bare
// URL, and the same shortening for a link that has no text of its own, so a
// workbook names an issue the way the app does.

// The http(s) prefix is load-bearing: it is what keeps javascript:/data:
// targets out of a cell's hyperlink. Labels are single-line; targets allow one
// nested parenthesis pair. Go's regexp is leftmost-first, so the Markdown
// alternative wins at the position where both could start.
var linkRE = regexp.MustCompile(`\[([^\]\n]*)\]\((https?://(?:[^\s()]|\([^\s()]*\))+)\)|https?://[^\s]+`)

type link struct {
	url   string
	label string
	// Whether the label was authored rather than derived from the URL.
	explicit bool
}

// extractLinks returns the distinct HTTP(S) links in a description, in
// first-occurrence order. An explicit label replaces a derived one for the same
// URL without moving it.
func extractLinks(text string) []link {
	var order []string
	seen := map[string]link{}
	for _, m := range linkRE.FindAllStringSubmatch(text, -1) {
		url, label, explicit := m[2], m[1], true
		if url == "" { // a bare URL: the whole match, minus sentence punctuation
			url, label, explicit = trimURL(m[0]), "", false
		}
		label = strings.TrimSpace(label)
		if label == "" {
			label, explicit = linkLabel(url), false
		}
		prev, ok := seen[url]
		if !ok {
			order = append(order, url)
		}
		if !ok || (explicit && !prev.explicit) {
			seen[url] = link{url: url, label: label, explicit: explicit}
		}
	}
	links := make([]link, 0, len(order))
	for _, url := range order {
		links = append(links, seen[url])
	}
	return links
}

// Known link shapes, recognised so that a link reads as the thing it points at
// rather than as a URL. The set is open and happens to hold one today; a shape
// is matched on the URL alone.
//
// This is presentation, not tracker integration. Deciding which of a
// description's links are the tracker's own issues is a separate job, needs the
// configured tracker host, and belongs to Recon — see meResponse.TrackerURL.
var issueKeyPath = regexp.MustCompile(`/browse/([A-Z0-9]+-\d+)`)

// linkLabel gives a link a short, readable name: an issue key when the URL is a
// ".../browse/KEY-123" link, otherwise the host plus the last path segment (so
// two links to the same host stay distinguishable).
//
// Parity with links.ts stops at a URL Go's parser rejects and a browser accepts
// — an invalid percent escape in the path. The cell then holds the whole URL
// where the app would shorten it.
func linkLabel(rawURL string) string {
	u, err := url.Parse(rawURL)
	if err != nil || u.Hostname() == "" {
		return rawURL
	}
	if m := issueKeyPath.FindStringSubmatch(u.EscapedPath()); m != nil {
		return m[1]
	}
	// Lowercased because URL.hostname is: a host is case-insensitive, and the
	// same link must not read two ways in the app and in the workbook.
	host := strings.ToLower(u.Hostname())
	segments := strings.FieldsFunc(u.EscapedPath(), func(r rune) bool { return r == '/' })
	if len(segments) == 0 {
		return host
	}
	return host + "/" + segments[len(segments)-1]
}

// trimURL strips sentence punctuation from a bare URL while preserving
// parentheses the URL itself balances.
func trimURL(url string) string {
	end := len(url)
	for end > 0 {
		ch := url[end-1]
		if !isTrailingPunct(ch) {
			break
		}
		if ch == ')' && countByte(url[:end], '(') >= countByte(url[:end], ')') {
			break
		}
		end--
	}
	return url[:end]
}

func isTrailingPunct(ch byte) bool {
	switch ch {
	case ')', ']', '.', ',', ';', '!', '?', '\'', '"':
		return true
	}
	return false
}

func countByte(s string, b byte) int {
	n := 0
	for i := 0; i < len(s); i++ {
		if s[i] == b {
			n++
		}
	}
	return n
}
